### PR TITLE
ci: asynchronous + only modified files

### LINKED ISSUE #95: improvement: CI/CD verifier script
To improve the cairo program verifier script, we could only test modified listings (git diff from master branch), or run tests asynchronously.

### PR DESCRIPTION
**Pull request by julienbrs**
**Issue:** closes #95 

### Description
This PR introduces optimizations to the CI by modifying the bash verifier script. Key changes include:

1. **Selective Testing**: The script now identifies and tests only the Cairo programs that have been modified.

2. **Asynchronous Execution**: The updated script executes tests in parallel.

3. **Error Tracking Across Processes**: A temporary file-based system is used to track errors across these asynchronous processes.

4. **Refined Logging and Error Handling**.

### CODE REVIEW

**Comment by julienbrs** [comment]
Struggled a bit, sorry for the flood of commits but should be good to merge now @julio4

File: scripts/cairo_programs_verifier.sh [hunk 1]
```diff
@@ -4,39 +4,67 @@ GREEN='\033[1;32m'
...
-        cat "$dir/error.log"
-      fi
-      rm "$dir/error.log"
+# root directory of the repository
+REPO_ROOT="$(git rev-parse --show-toplevel)"
+error_file=$(mktemp)
+
+# function to list modified cairo files
+list_modified_cairo_files() {
+    git diff --name-only main...HEAD -- listings | grep -E 'listings/ch.*/*.cairo$'
```
Comment (julio4) [inline scripts/cairo_programs_verifier.sh:13]: Less restrictive is ok, like `listings/ch.*/*`, so even if there's only a config change `Scarb.toml` the script still verify the listing.

File: scripts/cairo_programs_verifier.sh [hunk 2]
```diff
@@ -4,39 +4,67 @@ GREEN='\033[1;32m'
...
-  echo "\n${RED}Some projects have errors, please check the list above.${NC}\n"
-  exit 1
+if grep -q "1" "$error_file"; then
+    echo -e "\n${RED}Some projects have errors, please check the list above.${NC}\n"
+    rm "$error_file"
+    exit 1
 else
-  echo -e "\n${GREEN}All scarb builds were completed successfully${NC}.\n"
-  exit 0
+    echo -e "\n${GREEN}All scarb builds were completed successfully${NC}.\n"
```
Comment (julio4) [inline scripts/cairo_programs_verifier.sh:67]: Suggestion: count the number of detected changed listing and print `All {listings_count} buiilds were completed successfully` and if no change `No new changes detected.`

File: scripts/cairo_programs_verifier.sh (hunk 1 above)
Comment (julio4) [reply scripts/cairo_programs_verifier.sh:13]: Also we need to be sure to compare against the upstream `NethermindEth/StarknetByExample` main branch and not the local main branch.

**Comment by julienbrs** [comment]
@julio4 changes requested done!

**Comment by julio4** [comment]
@julienbrs Thank you for addressing the earlier changes.
I modified it a bit:
- Configure upstream when not set (for fork and clones)
- Fetch listings instead of changed files
- Only execute format and test if build step is passing
- Removed parallel executions: there's less listing to test so it's not too long to execute, this make error output better.

Also some changes in the github actions:
- PR: execute verify script on changes
- push to main: deploy mdbook without executing verify script

**Comment by julienbrs** [comment]
> @julienbrs Thank you for addressing the changes. I modified it a bit:
> 
>     * Configure upstream when not set (for fork and clones)
> 
>     * Fetch listings instead of modified files
> 
>     * Only execute format and test if build step is passing
> 
>     * Removed parallel executions: there's less listings to test so it's not too long to execute, this make error output better.
> 
> 
> Also some changes in the github actions:
> 
>     * PR: execute verify script (only on changed listings)
> 
>     * push to main: deploy mdbook without executing verify script
> 
> 
> Let me know if you agree with these changes

Your changes make sense, I agree with everything

**Pull request merged**